File: src/packages/job_queue/pg_queue.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Mapping, Sequence
import uuid

import sqlalchemy as sa
from sqlalchemy.dialects import postgresql
from sqlalchemy.ext.asyncio import AsyncSession

from packages.observability.context import new_trace_id, normalize_trace_id

from .protocol import (
    JOB_PAYLOAD_VERSION_V1,
    JOB_STATUS_DEAD,
    JOB_STATUS_DONE,
    JOB_STATUS_LEASED,
    JOB_STATUS_QUEUED,
    RUN_EXECUTE_JOB_TYPE,
    JobLease,
    JobLeaseLostError,
    JobQueue,
    default_retry_delay_seconds,
)

_metadata = sa.MetaData()

_jobs = sa.Table(
    "jobs",
    _metadata,
    sa.Column(
        "id",
        postgresql.UUID(as_uuid=True),
        primary_key=True,
        server_default=sa.text("gen_random_uuid()"),
    ),
    sa.Column("job_type", sa.Text(), nullable=False),
    sa.Column(
        "payload_json",
        postgresql.JSONB(astext_type=sa.Text()),
        nullable=False,
        server_default=sa.text("'{}'::jsonb"),
    ),
    sa.Column("status", sa.Text(), nullable=False, server_default=sa.text("'queued'")),
    sa.Column(
        "available_at",
        sa.TIMESTAMP(timezone=True),
        nullable=False,
        server_default=sa.text("now()"),
    ),
    sa.Column("leased_until", sa.TIMESTAMP(timezone=True), nullable=True),
    sa.Column("lease_token", postgresql.UUID(as_uuid=True), nullable=True),
    sa.Column("attempts", sa.Integer(), nullable=False, server_default=sa.text("0")),
    sa.Column(
        "created_at",
        sa.TIMESTAMP(timezone=True),
        nullable=False,
        server_default=sa.text("now()"),
    ),
    sa.Column(
        "updated_at",
        sa.TIMESTAMP(timezone=True),
        nullable=False,
        server_default=sa.text("now()"),
    ),
)


_LEASE_ATTEMPTS_REAP_LIMIT = 10


class SqlAlchemyPgJobQueue(JobQueue):
    def __init__(self, session: AsyncSession, *, max_attempts: int = 25) -> None:
        if max_attempts <= 0:
            raise ValueError("max_attempts 必须为正数")
        self._session = session
        self._max_attempts = max_attempts

    async def enqueue_run(
        self,
        *,
        org_id: uuid.UUID,
        run_id: uuid.UUID,
        trace_id: str | None,
        queue_job_type: str | None = None,
        payload: Mapping[str, Any] | None = None,
        available_at: datetime | None = None,
    ) -> uuid.UUID:
        job_id = uuid.uuid4()
        chosen_trace_id = normalize_trace_id(trace_id) or new_trace_id()
        chosen_job_type = (queue_job_type or "").strip() or RUN_EXECUTE_JOB_TYPE
        if not chosen_job_type:
            raise ValueError("queue_job_type 不能为空")
        # 约定：jobs.id 与 payload_json.job_id 必须一致，便于 worker 只看 payload 即可回放/审计。
        payload_json: dict[str, Any] = {
            "v": JOB_PAYLOAD_VERSION_V1,
            "job_id": str(job_id),
            "type": RUN_EXECUTE_JOB_TYPE,
            "trace_id": chosen_trace_id,
            "org_id": str(org_id),
            "run_id": str(run_id),
            "payload": dict(payload or {}),
        }
        values: dict[str, Any] = {
            "id": job_id,
            "job_type": chosen_job_type,
            "payload_json": payload_json,
            "status": JOB_STATUS_QUEUED,
            "leased_until": None,
            "attempts": 0,
            "updated_at": sa.func.now(),
        }
        if available_at is not None:
            values["available_at"] = available_at
        stmt = sa.insert(_jobs).values(values)
        await self._session.execute(stmt)
        return job_id

    async def lease(
        self,
        *,
        lease_seconds: int = 30,
        job_types: Sequence[str] | None = None,
    ) -> JobLease | None:
        if lease_seconds <= 0:
            raise ValueError("lease_seconds 必须为正数")

        chosen_job_types: tuple[str, ...] | None = None
        if job_types is not None:
            cleaned = [item.strip() for item in job_types if item and item.strip()]
            if not cleaned:
                return None
            seen: set[str] = set()
            deduped: list[str] = []
            for item in cleaned:
                if item in seen:
                    continue
                seen.add(item)
                deduped.append(item)
            chosen_job_types = tuple(deduped)

        for _ in range(_LEASE_ATTEMPTS_REAP_LIMIT):
            lease = await self._try_lease_one(lease_seconds=lease_seconds, job_types=chosen_job_types)
            if lease is not None:
                return lease
            if not await self._try_mark_dead_one(job_types=chosen_job_types):
                return None
        return None

    async def heartbeat(self, *, lease: JobLease, lease_seconds: int = 30) -> None:
        if lease_seconds <= 0:
            raise ValueError("lease_seconds 必须为正数")

        now = sa.func.now()
        lease_until = now + sa.func.make_interval(0, 0, 0, 0, 0, 0, lease_seconds)
        stmt = (
            sa.update(_jobs)
            .where(_jobs.c.id == lease.job_id)
            .where(_jobs.c.status == JOB_STATUS_LEASED)
            .where(_jobs.c.lease_token == lease.lease_token)
            .values(leased_until=lease_until, updated_at=now)
        )
        result = await self._session.execute(stmt)
        if result.rowcount != 1:
            raise JobLeaseLostError(job_id=lease.job_id)

    async def ack(self, *, lease: JobLease) -> None:
        stmt = (
            sa.update(_jobs)
            .where(_jobs.c.id == lease.job_id)
            .where(_jobs.c.status == JOB_STATUS_LEASED)
            .where(_jobs.c.lease_token == lease.lease_token)
            .values(
                status=JOB_STATUS_DONE,
                leased_until=None,
                lease_token=None,
                updated_at=sa.func.now(),
            )
        )
        result = await self._session.execute(stmt)
        if result.rowcount != 1:
            raise JobLeaseLostError(job_id=lease.job_id)

    async def nack(self, *, lease: JobLease, delay_seconds: int | None = None) -> None:
        if lease.attempts >= self._max_attempts:
            await self._dead_letter(lease=lease)
            return

        chosen_delay = delay_seconds
        if chosen_delay is None:
            chosen_delay = default_retry_delay_seconds(attempts=lease.attempts)
        if chosen_delay < 0:
            raise ValueError("delay_seconds 不能为负数")

        now = sa.func.now()
        retry_at = now + sa.func.make_interval(0, 0, 0, 0, 0, 0, chosen_delay)
        stmt = (
            sa.update(_jobs)
            .where(_jobs.c.id == lease.job_id)
            .where(_jobs.c.status == JOB_STATUS_LEASED)
            .where(_jobs.c.lease_token == lease.lease_token)
            .values(
                status=JOB_STATUS_QUEUED,
                leased_until=None,
                lease_token=None,
                available_at=retry_at,
                updated_at=now,
            )
        )
        result = await self._session.execute(stmt)
        if result.rowcount != 1:
            raise JobLeaseLostError(job_id=lease.job_id)

    async def _try_lease_one(
        self,
        *,
        lease_seconds: int,
        job_types: tuple[str, ...] | None,
    ) -> JobLease | None:
        now = sa.func.now()
        lease_until = now + sa.func.make_interval(0, 0, 0, 0, 0, 0, lease_seconds)
        candidate = self._candidate_job_cte(now=now, attempts_lt=self._max_attempts, job_types=job_types)
        stmt = (
            sa.update(_jobs)
            .where(_jobs.c.id == candidate.c.id)
            .values(
                status=JOB_STATUS_LEASED,
                leased_until=lease_until,
                lease_token=sa.func.gen_random_uuid(),
                attempts=_jobs.c.attempts + 1,
                updated_at=now,
            )
            .returning(
                _jobs.c.id.label("job_id"),
                _jobs.c.job_type,
                _jobs.c.payload_json,
                _jobs.c.attempts,
                _jobs.c.leased_until,
                _jobs.c.lease_token,
            )
        )
        row = (await self._session.execute(stmt)).mappings().one_or_none()
        if row is None:
            return None
        return JobLease(
            job_id=row["job_id"],
            job_type=row["job_type"],
            payload_json=row["payload_json"],
            attempts=int(row["attempts"]),
            leased_until=row["leased_until"],
            lease_token=row["lease_token"],
        )

    async def _try_mark_dead_one(self, *, job_types: tuple[str, ...] | None) -> bool:
        now = sa.func.now()
        candidate = self._candidate_job_cte(now=now, attempts_gte=self._max_attempts, job_types=job_types)
        stmt = (
            sa.update(_jobs)
            .where(_jobs.c.id == candidate.c.id)
            .values(
                status=JOB_STATUS_DEAD,
                leased_until=None,
                lease_token=None,
                updated_at=now,
            )
        )
        result = await self._session.execute(stmt)
        return bool(result.rowcount == 1)

    async def _dead_letter(self, *, lease: JobLease) -> None:
        stmt = (
            sa.update(_jobs)
            .where(_jobs.c.id == lease.job_id)
            .where(_jobs.c.status == JOB_STATUS_LEASED)
            .where(_jobs.c.lease_token == lease.lease_token)
            .values(
                status=JOB_STATUS_DEAD,
                leased_until=None,
                lease_token=None,
                updated_at=sa.func.now(),
            )
        )
        result = await self._session.execute(stmt)
        if result.rowcount != 1:
            raise JobLeaseLostError(job_id=lease.job_id)

    def _candidate_job_cte(
        self,
        *,
        now: sa.SQLColumnExpression[Any],
        attempts_lt: int | None = None,
        attempts_gte: int | None = None,
        job_types: tuple[str, ...] | None = None,
    ) -> sa.CTE:
        if attempts_lt is not None and attempts_gte is not None:
            raise ValueError("attempts_lt 与 attempts_gte 不能同时设置")

        conditions: list[sa.ColumnElement[bool]] = [
            sa.or_(
                sa.and_(
                    _jobs.c.status == JOB_STATUS_QUEUED,
                    _jobs.c.available_at <= now,
                ),
                sa.and_(
                    _jobs.c.status == JOB_STATUS_LEASED,
                    _jobs.c.leased_until.is_not(None),
                    _jobs.c.leased_until <= now,
                ),
            )
        ]
        if attempts_lt is not None:
            conditions.append(_jobs.c.attempts < attempts_lt)
        if attempts_gte is not None:
            conditions.append(_jobs.c.attempts >= attempts_gte)
        if job_types is not None:
            conditions.append(_jobs.c.job_type.in_(job_types))

        return (
            sa.select(_jobs.c.id)
            .where(sa.and_(*conditions))
            .order_by(_jobs.c.available_at.asc(), _jobs.c.created_at.asc(), _jobs.c.id.asc())
            .with_for_update(skip_locked=True)
            .limit(1)
            .cte("candidate")
        )


__all__ = ["SqlAlchemyPgJobQueue"]

File: src/services/worker/internal/queue/pg_queue.go
//go:build !desktop

package queue

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"
)

const activeRunExecuteJobIndex = "ux_jobs_run_execute_active_run"

type PgQueue struct {
	pool         *pgxpool.Pool
	maxAttempts  int
	capabilities []string
}

func NewPgQueue(pool *pgxpool.Pool, maxAttempts int, capabilities []string) (*PgQueue, error) {
	if pool == nil {
		return nil, fmt.Errorf("pool must not be nil")
	}
	if maxAttempts <= 0 {
		return nil, fmt.Errorf("max_attempts must be positive")
	}
	caps := capabilities
	if caps == nil {
		caps = []string{}
	}
	return &PgQueue{pool: pool, maxAttempts: maxAttempts, capabilities: caps}, nil
}

func (q *PgQueue) EnqueueRun(
	ctx context.Context,
	accountID uuid.UUID,
	runID uuid.UUID,
	traceID string,
	queueJobType string,
	payload map[string]any,
	availableAt *time.Time,
) (uuid.UUID, error) {
	jobID := uuid.New()
	chosenTraceID := normalizeTraceID(traceID)
	if chosenTraceID == "" {
		chosenTraceID = uuid.New().String()
		chosenTraceID = strings.ReplaceAll(chosenTraceID, "-", "")
	}
	chosenJobType := strings.TrimSpace(queueJobType)
	if chosenJobType == "" {
		chosenJobType = RunExecuteJobType
	}

	payloadCopy := map[string]any{}
	for key, value := range payload {
		payloadCopy[key] = value
	}

	payloadJSON := map[string]any{
		"v":          JobPayloadVersionV1,
		"job_id":     jobID.String(),
		"type":       chosenJobType,
		"trace_id":   chosenTraceID,
		"account_id": accountID.String(),
		"run_id":     runID.String(),
		"payload":    payloadCopy,
	}

	encoded, err := json.Marshal(payloadJSON)
	if err != nil {
		return uuid.Nil, err
	}

	if chosenJobType == RunExecuteJobType {
		existingJobID, err := q.findActiveRunExecuteJob(ctx, runID)
		if err != nil {
			return uuid.Nil, err
		}
		if existingJobID != uuid.Nil {
			return uuid.Nil, fmt.Errorf("%w: run_id=%s job_id=%s", ErrRunExecuteAlreadyQueued, runID, existingJobID)
		}
	}

	_, err = q.pool.Exec(
		ctx,
		`INSERT INTO jobs (
			id, job_type, payload_json, status, available_at,
			leased_until, lease_token, attempts, created_at, updated_at
		) VALUES (
			$1, $2, $3::jsonb, $4, COALESCE($5, now()),
			NULL, NULL, 0, now(), now()
		)`,
		jobID,
		chosenJobType,
		string(encoded),
		JobStatusQueued,
		availableAt,
	)
	if err != nil {
		if chosenJobType == RunExecuteJobType && isActiveRunExecuteConflict(err) {
			existingJobID, lookupErr := q.findActiveRunExecuteJob(ctx, runID)
			if lookupErr == nil && existingJobID != uuid.Nil {
				return uuid.Nil, fmt.Errorf("%w: run_id=%s job_id=%s", ErrRunExecuteAlreadyQueued, runID, existingJobID)
			}
			// 即使找不到（对方 job 瞬间完成），unique 冲突本身已经明确，仍然返回语义错误
			return uuid.Nil, fmt.Errorf("%w: run_id=%s", ErrRunExecuteAlreadyQueued, runID)
		}
		return uuid.Nil, err
	}

	_, _ = q.pool.Exec(ctx, `SELECT pg_notify('arkloop:jobs', '')`)

	return jobID, nil
}

func isActiveRunExecuteConflict(err error) bool {
	var pgErr *pgconn.PgError
	return errors.As(err, &pgErr) && pgErr.Code == "23505" && pgErr.ConstraintName == activeRunExecuteJobIndex
}

func (q *PgQueue) findActiveRunExecuteJob(ctx context.Context, runID uuid.UUID) (uuid.UUID, error) {
	if runID == uuid.Nil {
		return uuid.Nil, nil
	}
	var existingJobID uuid.UUID
	err := q.pool.QueryRow(
		ctx,
		`SELECT id
		   FROM jobs
		  WHERE job_type = $1
		    AND payload_json->>'run_id' = $2
		    AND status IN ($3, $4)
		  ORDER BY created_at ASC, id ASC
		  LIMIT 1`,
		RunExecuteJobType,
		runID.String(),
		JobStatusQueued,
		JobStatusLeased,
	).Scan(&existingJobID)
	if errors.Is(err, pgx.ErrNoRows) {
		return uuid.Nil, nil
	}
	if err != nil {
		return uuid.Nil, err
	}
	return existingJobID, nil
}

func (q *PgQueue) Lease(ctx context.Context, leaseSeconds int, jobTypes []string) (*JobLease, error) {
	if leaseSeconds <= 0 {
		return nil, fmt.Errorf("lease_seconds must be positive")
	}

	chosenJobTypes := normalizeJobTypes(jobTypes)
	if len(chosenJobTypes) == 0 {
		return nil, nil
	}

	for i := 0; i < leaseAttemptsReapLimit; i++ {
		lease, err := q.tryLeaseOne(ctx, leaseSeconds, chosenJobTypes)
		if err != nil {
			return nil, err
		}
		if lease != nil {
			return lease, nil
		}

		marked, err := q.tryMarkDeadOne(ctx, chosenJobTypes)
		if err != nil {
			return nil, err
		}
		if !marked {
			return nil, nil
		}
	}

	return nil, nil
}

func (q *PgQueue) Heartbeat(ctx context.Context, lease JobLease, leaseSeconds int) error {
	if leaseSeconds <= 0 {
		return fmt.Errorf("lease_seconds must be positive")
	}

	result, err := q.pool.Exec(
		ctx,
		`UPDATE jobs
		 SET leased_until = now() + ($1 * interval '1 second'),
		     updated_at = now()
		 WHERE id = $2
		   AND status = $3
		   AND lease_token = $4`,
		leaseSeconds,
		lease.JobID,
		JobStatusLeased,
		lease.LeaseToken,
	)
	if err != nil {
		return err
	}
	if result.RowsAffected() != 1 {
		return &LeaseLostError{JobID: lease.JobID}
	}
	return nil
}

func (q *PgQueue) Ack(ctx context.Context, lease JobLease) error {
	result, err := q.pool.Exec(
		ctx,
		`UPDATE jobs
		 SET status = $1,
		     leased_until = NULL,
		     lease_token = NULL,
		     updated_at = now()
		 WHERE id = $2
		   AND status = $3
		   AND lease_token = $4`,
		JobStatusDone,
		lease.JobID,
		JobStatusLeased,
		lease.LeaseToken,
	)
	if err != nil {
		return err
	}
	if result.RowsAffected() != 1 {
		return &LeaseLostError{JobID: lease.JobID}
	}
	return nil
}

func (q *PgQueue) Nack(ctx context.Context, lease JobLease, delaySeconds *int) error {
	if lease.Attempts >= q.maxAttempts {
		return q.deadLetter(ctx, lease)
	}

	chosenDelay := DefaultRetryDelaySeconds(lease.Attempts)
	if delaySeconds != nil {
		chosenDelay = *delaySeconds
	}
	if chosenDelay < 0 {
		return fmt.Errorf("delay_seconds must not be negative")
	}

	result, err := q.pool.Exec(
		ctx,
		`UPDATE jobs
		 SET status = $1,
		     leased_until = NULL,
		     lease_token = NULL,
		     available_at = now() + ($2 * interval '1 second'),
		     updated_at = now()
		 WHERE id = $3
		   AND status = $4
		   AND lease_token = $5`,
		JobStatusQueued,
		chosenDelay,
		lease.JobID,
		JobStatusLeased,
		lease.LeaseToken,
	)
	if err != nil {
		return err
	}
	if result.RowsAffected() != 1 {
		return &LeaseLostError{JobID: lease.JobID}
	}
	return nil
}

func (q *PgQueue) tryLeaseOne(ctx context.Context, leaseSeconds int, jobTypes []string) (*JobLease, error) {
	tx, err := q.pool.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return nil, err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	var (
		jobID       uuid.UUID
		jobType     string
		payloadRaw  []byte
		currentTrys int
	)

	err = tx.QueryRow(
		ctx,
		`SELECT id, job_type, payload_json, attempts
		 FROM jobs
		 WHERE (
		   (status = $1 AND available_at <= now()) OR
		   (status = $2 AND leased_until IS NOT NULL AND leased_until <= now())
		 )
		 AND attempts < $3
		 AND job_type = ANY($4)
		 AND worker_tags <@ $5
		 ORDER BY available_at ASC, created_at ASC, id ASC
		 FOR UPDATE SKIP LOCKED
		 LIMIT 1`,
		JobStatusQueued,
		JobStatusLeased,
		q.maxAttempts,
		jobTypes,
		q.capabilities,
	).Scan(&jobID, &jobType, &payloadRaw, &currentTrys)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, nil
		}
		return nil, err
	}

	leaseToken := uuid.New()
	var (
		attempts    int
		leasedUntil time.Time
	)
	err = tx.QueryRow(
		ctx,
		`UPDATE jobs
		 SET status = $1,
		     leased_until = now() + ($2 * interval '1 second'),
		     lease_token = $3,
		     attempts = attempts + 1,
		     updated_at = now()
		 WHERE id = $4
		 RETURNING attempts, leased_until`,
		JobStatusLeased,
		leaseSeconds,
		leaseToken,
		jobID,
	).Scan(&attempts, &leasedUntil)
	if err != nil {
		return nil, err
	}

	if err := tx.Commit(ctx); err != nil {
		return nil, err
	}

	payloadJSON := map[string]any{}
	if len(payloadRaw) > 0 {
		if err := json.Unmarshal(payloadRaw, &payloadJSON); err != nil {
			return nil, err
		}
	}

	return &JobLease{
		JobID:       jobID,
		JobType:     jobType,
		PayloadJSON: payloadJSON,
		Attempts:    attempts,
		LeasedUntil: leasedUntil,
		LeaseToken:  leaseToken,
	}, nil
}

func (q *PgQueue) tryMarkDeadOne(ctx context.Context, jobTypes []string) (bool, error) {
	tx, err := q.pool.BeginTx(ctx, pgx.TxOptions{})
	if err != nil {
		return false, err
	}
	defer func() { _ = tx.Rollback(ctx) }()

	var jobID uuid.UUID
	err = tx.QueryRow(
		ctx,
		`SELECT id
		 FROM jobs
		 WHERE (
		   (status = $1 AND available_at <= now()) OR
		   (status = $2 AND leased_until IS NOT NULL AND leased_until <= now())
		 )
		 AND attempts >= $3
		 AND job_type = ANY($4)
		 AND worker_tags <@ $5
		 ORDER BY available_at ASC, created_at ASC, id ASC
		 FOR UPDATE SKIP LOCKED
		 LIMIT 1`,
		JobStatusQueued,
		JobStatusLeased,
		q.maxAttempts,
		jobTypes,
		q.capabilities,
	).Scan(&jobID)
	if err != nil {
		if err == pgx.ErrNoRows {
			return false, nil
		}
		return false, err
	}

	if _, err := tx.Exec(
		ctx,
		`UPDATE jobs
		 SET status = $1,
		     leased_until = NULL,
		     lease_token = NULL,
		     updated_at = now()
		 WHERE id = $2`,
		JobStatusDead,
		jobID,
	); err != nil {
		return false, err
	}

	if err := tx.Commit(ctx); err != nil {
		return false, err
	}
	return true, nil
}

func (q *PgQueue) QueueDepth(ctx context.Context, jobTypes []string) (int, error) {
	chosenJobTypes := normalizeJobTypes(jobTypes)
	if len(chosenJobTypes) == 0 {
		return 0, nil
	}

	var count int
	err := q.pool.QueryRow(ctx,
		`SELECT COUNT(*)
		 FROM jobs
		 WHERE status = $1
		   AND available_at <= now()
		   AND job_type = ANY($2)
		   AND worker_tags <@ $3`,
		JobStatusQueued,
		chosenJobTypes,
		q.capabilities,
	).Scan(&count)
	if err != nil {
		return 0, err
	}
	return count, nil
}

func (q *PgQueue) QueueStats(ctx context.Context, jobTypes []string) (QueueStats, error) {
	chosenJobTypes := normalizeJobTypes(jobTypes)
	if len(chosenJobTypes) == 0 {
		return QueueStats{}, nil
	}

	var stats QueueStats
	var oldestReadyAgeSeconds int64
	err := q.pool.QueryRow(ctx,
		`SELECT
		   COUNT(*) FILTER (WHERE status = $1 AND available_at <= now()),
		   COUNT(*) FILTER (WHERE status = $2 AND leased_until > now()),
		   COALESCE(
		     EXTRACT(EPOCH FROM (
		       now() - MIN(available_at) FILTER (
		         WHERE status = $1 AND available_at <= now()
		       )
		     ))::bigint,
		     0
		   )
		 FROM jobs
		 WHERE job_type = ANY($3)
		   AND worker_tags <@ $4`,
		JobStatusQueued,
		JobStatusLeased,
		chosenJobTypes,
		q.capabilities,
	).Scan(&stats.ReadyDepth, &stats.InFlight, &oldestReadyAgeSeconds)
	if err != nil {
		return QueueStats{}, err
	}
	stats.OldestReadyAge = time.Duration(oldestReadyAgeSeconds) * time.Second
	return stats, nil
}

func (q *PgQueue) deadLetter(ctx context.Context, lease JobLease) error {
	result, err := q.pool.Exec(
		ctx,
		`UPDATE jobs
		 SET status = $1,
		     leased_until = NULL,
		     lease_token = NULL,
		     updated_at = now()
		 WHERE id = $2
		   AND status = $3
		   AND lease_token = $4`,
		JobStatusDead,
		lease.JobID,
		JobStatusLeased,
		lease.LeaseToken,
	)
	if err != nil {
		return err
	}
	if result.RowsAffected() != 1 {
		return &LeaseLostError{JobID: lease.JobID}
	}
	return nil
}
